In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')

In [2]:
data = pd.read_csv('HW1-data.csv')
data.head()

User  Gender (1 =F, 0=M)  260: Star Wars: Episode IV - A New Hope (1977)  \
0   755                   0                                             1.0   
1  5277                   0                                             5.0   
2  1577                   1                                             NaN   
3  4388                   0                                             NaN   
4  1202                   1                                             4.0   

   1210: Star Wars: Episode VI - Return of the Jedi (1983)  \
0                                                5.0         
1                                                3.0         
2                                                NaN         
3                                                3.0         
4                                                3.0         

   356: Forrest Gump (1994)  318: Shawshank Redemption, The (1994)  \
0                       2.0                                    NaN   
1                       NaN                                    2.0   
2                       NaN                                    5.0   
3                       NaN                                    NaN   
4                       4.0                                    1.0   

   593: Silence of the Lambs, The (1991)  3578: Gladiator (2000)  \
0                                    4.0                     4.0   
1                                    4.0                     2.0   
2                                    2.0                     NaN   
3                                    NaN                     1.0   
4                                    4.0                     1.0   

   1: Toy Story (1995)  2028: Saving Private Ryan (1998)  ...  \
0                  2.0                               2.0  ...   
1                  1.0                               NaN  ...   
2                  4.0                               NaN  ...   
3                  2.0                               3.0  ...   
4                  NaN                               4.0  ...   

   2396: Shakespeare in Love (1998)  2916: Total Recall (1990)  \
0                               2.0                        NaN   
1                               3.0                        2.0   
2                               NaN                        1.0   
3                               NaN                        4.0   
4                               5.0                        1.0   

   780: Independence Day (ID4) (1996)  541: Blade Runner (1982)  \
0                                 5.0                       2.0   
1                                 2.0                       NaN   
2                                 4.0                       4.0   
3                                 1.0                       3.0   
4                                 NaN                       4.0   

   1265: Groundhog Day (1993)  2571: Matrix, The (1999)  \
0                         NaN                       4.0   
1                         2.0                       NaN   
2                         1.0                       1.0   
3                         5.0                       NaN   
4                         NaN                       3.0   

   527: Schindler's List (1993)  2762: Sixth Sense, The (1999)  \
0                           2.0                            5.0   
1                           5.0                            1.0   
2                           2.0                            3.0   
3                           5.0                            1.0   
4                           5.0                            5.0   

   1198: Raiders of the Lost Ark (1981)  34: Babe (1995)  
0                                   NaN              NaN  
1                                   3.0              NaN  
2                                   1.0              3.0  
3                                   1.0              2.0  
4                                   NaN              NaN  

[5 rows x 22 columns]

In [3]:
data.shape

(20, 22)

## non-personalized summary statistics
1. Mean Rating: Calculate the mean rating for each movie, order with the highest
rating listed first, and submit the top three (along with the mean scores for the top
two).

In [4]:
# by default, pandas exclude NAs
df_ratings = data[data.columns[~data.columns.isin(['User','Gender (1 =F, 0=M)'])]]
df_ratings.mean().sort_values(ascending = False)[0:3]

318: Shawshank Redemption, The (1994)             3.600000
260: Star Wars: Episode IV - A New Hope (1977)    3.266667
541: Blade Runner (1982)                          3.222222
dtype: float64

2. Rating Count (popularity): Count the number of ratings for each movie, order with the most number of ratings first, and submit the top three (along with the counts for the top two). 

In [5]:
df_ratings.count().sort_values(ascending = False)[0:3]

1: Toy Story (1995)                               17
593: Silence of the Lambs, The (1991)             16
260: Star Wars: Episode IV - A New Hope (1977)    15
dtype: int64

3. % of ratings 4+ (liking): Calculate the percentage of ratings for each movie that are 4 or higher. Order with the highest percentage first, and submit the top three (along with the percentage for the top two).  Notice that the three different measures of "best" reflect different priorities and give different results; this should help you see why you need to be thoughtful about what metrics you use. 

In [6]:
df_ratings.T >= 4

0      1      2  \
260: Star Wars: Episode IV - A New Hope (1977)      False   True  False   
1210: Star Wars: Episode VI - Return of the Jed...   True  False  False   
356: Forrest Gump (1994)                            False  False  False   
318: Shawshank Redemption, The (1994)               False  False   True   
593: Silence of the Lambs, The (1991)                True   True  False   
3578: Gladiator (2000)                               True  False  False   
1: Toy Story (1995)                                 False  False   True   
2028: Saving Private Ryan (1998)                    False  False  False   
296: Pulp Fiction (1994)                            False  False  False   
1259: Stand by Me (1986)                            False   True  False   
2396: Shakespeare in Love (1998)                    False  False  False   
2916: Total Recall (1990)                           False  False  False   
780: Independence Day (ID4) (1996)                   True  False   True   
541: Blade Runner (1982)                            False  False   True   
1265: Groundhog Day (1993)                          False  False  False   
2571: Matrix, The (1999)                             True  False  False   
527: Schindler's List (1993)                        False   True  False   
2762: Sixth Sense, The (1999)                        True  False  False   
1198: Raiders of the Lost Ark (1981)                False  False  False   
34: Babe (1995)                                     False  False  False   

                                                        3      4      5  \
260: Star Wars: Episode IV - A New Hope (1977)      False   True  False   
1210: Star Wars: Episode VI - Return of the Jed...  False  False   True   
356: Forrest Gump (1994)                            False   True   True   
318: Shawshank Redemption, The (1994)               False  False   True   
593: Silence of the Lambs, The (1991)               False   True  False   
3578: Gladiator (2000)                              False  False  False   
1: Toy Story (1995)                                 False  False  False   
2028: Saving Private Ryan (1998)                    False   True  False   
296: Pulp Fiction (1994)                             True  False   True   
1259: Stand by Me (1986)                            False  False   True   
2396: Shakespeare in Love (1998)                    False   True   True   
2916: Total Recall (1990)                            True  False  False   
780: Independence Day (ID4) (1996)                  False  False   True   
541: Blade Runner (1982)                            False   True  False   
1265: Groundhog Day (1993)                           True  False  False   
2571: Matrix, The (1999)                            False  False  False   
527: Schindler's List (1993)                         True   True  False   
2762: Sixth Sense, The (1999)                       False   True  False   
1198: Raiders of the Lost Ark (1981)                False  False  False   
34: Babe (1995)                                     False  False  False   

                                                        6      7      8  \
260: Star Wars: Episode IV - A New Hope (1977)      False   True   True   
1210: Star Wars: Episode VI - Return of the Jed...  False  False  False   
356: Forrest Gump (1994)                            False  False  False   
318: Shawshank Redemption, The (1994)               False  False   True   
593: Silence of the Lambs, The (1991)               False  False  False   
3578: Gladiator (2000)                               True  False   True   
1: Toy Story (1995)                                 False  False   True   
2028: Saving Private Ryan (1998)                     True  False   True   
296: Pulp Fiction (1994)                            False  False  False   
1259: Stand by Me (1986)                            False  False  False   
2396: Shakespeare in Love (1998)                    False  False  False   
2916

In [7]:
(df_ratings >= 4).sum()

260: Star Wars: Episode IV - A New Hope (1977)             8
1210: Star Wars: Episode VI - Return of the Jedi (1983)    5
356: Forrest Gump (1994)                                   3
318: Shawshank Redemption, The (1994)                      7
593: Silence of the Lambs, The (1991)                      7
3578: Gladiator (2000)                                     6
1: Toy Story (1995)                                        6
2028: Saving Private Ryan (1998)                           4
296: Pulp Fiction (1994)                                   4
1259: Stand by Me (1986)                                   4
2396: Shakespeare in Love (1998)                           3
2916: Total Recall (1990)                                  1
780: Independence Day (ID4) (1996)                         4
541: Blade Runner (1982)                                   4
1265: Groundhog Day (1993)                                 5
2571: Matrix, The (1999)                                   5
527: Schindler's List (1

In [8]:
((df_ratings >= 4).sum()/(df_ratings >= 0).sum()).sort_values(ascending = False)[0:3]

318: Shawshank Redemption, The (1994)             0.700000
260: Star Wars: Episode IV - A New Hope (1977)    0.533333
3578: Gladiator (2000)                            0.500000
dtype: float64

##  product association 
using two different mechanisms (we aren't using lift here because of how similar the popularity levels are).

4. Top movies for someone who has seen Toy Story: Calculate movies that most often occur with Movie #1:  Toy Story, using the (x+y)/x method described in class. In other words, for each movie, calculate the percentage of Toy Story raters who also rated that movie. Order with the highest percentage first, and submit the top 3 (along with the correlations for the top two).  Note, you will have ties -- to break the ties, use the lowest-numbered movie as the higher-ranked one.  In other words, if Movies 541 and 318 are tied, then 318 gets the higher rank

In [13]:
count_associated_movie = (df_ratings[df_ratings['1: Toy Story (1995)']>=0]>=0).sum().sort_values(ascending = False).to_frame(name = 'associated_count')
count_associated_movie[0:8]

associated_count
1: Toy Story (1995)                                               17
260: Star Wars: Episode IV - A New Hope (1977)                    14
1210: Star Wars: Episode VI - Return of the Jed...                13
593: Silence of the Lambs, The (1991)                             13
780: Independence Day (ID4) (1996)                                13
1265: Groundhog Day (1993)                                        11
2916: Total Recall (1990)                                         11
296: Pulp Fiction (1994)                                          10

In [10]:
total_movie_counts = df_ratings.count().sort_values(ascending = False).to_frame(name = 'total_count')
total_movie_counts[0:4]

total_count
1: Toy Story (1995)                                          17
593: Silence of the Lambs, The (1991)                        16
260: Star Wars: Episode IV - A New Hope (1977)               15
1210: Star Wars: Episode VI - Return of the Jed...           14

In [11]:
df_toy = count_associated_movie.join( total_movie_counts)
df_toy['associated_perc'] = df_toy['associated_count']/df_toy['total_count']
df_toy = df_toy.sort_values(by = 'associated_perc', ascending = False)
df_toy.head(5)

associated_count  \
1: Toy Story (1995)                                               17   
780: Independence Day (ID4) (1996)                                13   
260: Star Wars: Episode IV - A New Hope (1977)                    14   
1210: Star Wars: Episode VI - Return of the Jed...                13   
1265: Groundhog Day (1993)                                        11   

                                                    total_count  \
1: Toy Story (1995)                                          17   
780: Independence Day (ID4) (1996)                           13   
260: Star Wars: Episode IV - A New Hope (1977)               15   
1210: Star Wars: Episode VI - Return of the Jed...           14   
1265: Groundhog Day (1993)                                   12   

                                                    associated_perc  
1: Toy Story (1995)                                        1.000000  
780: Independence Day (ID4) (1996)                         1.000000  
260: Star Wars: Episode IV - A New Hope (1977)             0.933333  
1210: Star Wars: Episode VI - Return of the Jed...         0.928571  
1265: Groundhog Day (1993)                                 0.916667

associated_count  total_count  \
3578: Gladiator (2000)                           9           12   
2571: Matrix, The (1999)                         9           12   
2396: Shakespeare in Love (1998)                 8           11   
2028: Saving Private Ryan (1998)                 8           11   
356: Forrest Gump (1994)                         7           10   

                                  associated_perc  
3578: Gladiator (2000)                   0.750000  
2571: Matrix, The (1999)                 0.750000  
2396: Shakespeare in Love (1998)         0.727273  
2028: Saving Private Ryan (1998)         0.727273  
356: Forrest Gump (1994)                 0.700000

5. Correlation with Toy Story:  Calculate the correlation between the vectors of ratings for Toy Story and each other movie.  You can use the built-in CORREL() function.  Order by the highest correlation (positive only) and submit the top 3 along with the correlation values for the top 2.  Notice the differences between co-occurrence and correlation; these metrics are showing different types of relationships.

In [18]:
df_ratings.corr()["1: Toy Story (1995)"].sort_values(ascending = False)

1: Toy Story (1995)                                        1.000000
318: Shawshank Redemption, The (1994)                      0.888523
34: Babe (1995)                                            0.811107
296: Pulp Fiction (1994)                                   0.709842
2028: Saving Private Ryan (1998)                           0.596849
356: Forrest Gump (1994)                                   0.522913
541: Blade Runner (1982)                                   0.329634
3578: Gladiator (2000)                                     0.253396
2916: Total Recall (1990)                                  0.163796
2396: Shakespeare in Love (1998)                           0.101768
1265: Groundhog Day (1993)                                -0.062858
780: Independence Day (ID4) (1996)                        -0.069923
260: Star Wars: Episode IV - A New Hope (1977)            -0.119005
527: Schindler's List (1993)                              -0.220315
2762: Sixth Sense, The (1999)                   

## demographic analysis 

6. Mean rating difference by gender:  First, recompute the mean rating for each movie separately for males and for females.  And calculate the overall mean rating (across all ratings) for males and females. Submit the two movies that have the greatest differences (one where men are most above women, and one where women are most above men) along with the differences in average.  Also submit the difference in overall rating averages (female average - male average).

In [35]:
df_gender = data[data.columns[~data.columns.isin(['User'])]].groupby(["Gender (1 =F, 0=M)"]).mean().T
df_gender['diff'] = df_gender[1] - df_gender[0] #female - male
df_gender = df_gender.sort_values(by = 'diff', ascending = False)
df_gender

Gender (1 =F, 0=M)                                         0         1  \
2396: Shakespeare in Love (1998)                    2.142857  4.250000   
34: Babe (1995)                                     2.000000  3.428571   
296: Pulp Fiction (1994)                            2.625000  4.000000   
1: Toy Story (1995)                                 2.300000  3.571429   
356: Forrest Gump (1994)                            2.250000  3.000000   
541: Blade Runner (1982)                            3.000000  3.500000   
318: Shawshank Redemption, The (1994)               3.400000  3.800000   
2762: Sixth Sense, The (1999)                       2.666667  3.000000   
260: Star Wars: Episode IV - A New Hope (1977)      3.125000  3.428571   
3578: Gladiator (2000)                              2.833333  3.000000   
1210: Star Wars: Episode VI - Return of the Jed...  3.000000  3.000000   
780: Independence Day (ID4) (1996)                  2.857143  2.666667   
2028: Saving Private Ryan (1998)                    3.142857  2.750000   
2916: Total Recall (1990)                           2.200000  1.714286   
1259: Stand by Me (1986)                            3.000000  2.428571   
593: Silence of the Lambs, The (1991)               3.333333  2.714286   
1265: Groundhog Day (1993)                          3.500000  2.833333   
2571: Matrix, The (1999)                            3.142857  2.400000   
527: Schindler's List (1993)                        3.500000  2.500000   
1198: Raiders of the Lost Ark (1981)                3.666667  2.000000   

Gender (1 =F, 0=M)                                      diff  
2396: Shakespeare in Love (1998)                    2.107143  
34: Babe (1995)                                     1.428571  
296: Pulp Fiction (1994)                            1.375000  
1: Toy Story (1995)                                 1.271429  
356: Forrest Gump (1994)                            0.750000  
541: Blade Runner (1982)                            0.500000  
318: Shawshank Redemption, The (1994)               0.400000  
2762: Sixth Sense, The (1999)                       0.333333  
260: Star Wars: Episode IV - A New Hope (1977)      0.303571  
3578: Gladiator (2000)                              0.166667  
1210: Star Wars: Episode VI - Return of the Jed...  0.000000  
780: Independence Day (ID4) (1996)                 -0.190476  
2028: Saving Private Ryan (1998)                   -0.392857  
2916: Total Recall (1990)                          -0.485714  
1259: Stand by Me (1986)                           -0.571429  
593: Silence of the Lambs, The (1991)              -0.619048  
1265: Groundhog Day (1993)                         -0.666667  
2571: Matrix, The (1999)                           -0.742857  
527: Schindler's List (1993)                       -1.000000  
1198: Raiders of the Lost Ark (1981)               -1.666667

In [36]:
df_gender.describe()

Gender (1 =F, 0=M)          0          1       diff
count               20.000000  20.000000  20.000000
mean                 2.884286   2.999286   0.115000
std                  0.499016   0.649655   0.941371
min                  2.000000   1.714286  -1.666667
25%                  2.543750   2.625000  -0.583333
50%                  3.000000   3.000000   0.083333
75%                  3.190476   3.446429   0.562500
max                  3.666667   4.250000   2.107143

7. Next, compute the % of ratings 4+ separately for males and females.  You'll be asked to submit two movies as above (largest difference in each direction). And again you'll indicate whether men or women are more likely to rate movies 4 stars or above.

In [68]:
df_rating_gender = pd.DataFrame()
for gender in (0, 1):
    df_gender = data[data["Gender (1 =F, 0=M)"]== gender]
    df_rating = df_gender[df_gender.columns[~df_gender.columns.isin(['Gender (1 =F, 0=M)', 'User'])]]
    df_rating_gender_temp = ((df_rating >= 4).sum()/(df_rating >= 0).sum()).sort_values(ascending = False).to_frame(name = "rating")
    df_rating_gender_temp.reset_index(level=0, inplace=True)
    df_rating_gender_temp['gender'] = gender
#    print(df_rating_gender_temp.head())
    df_rating_gender = pd.concat([df_rating_gender,df_rating_gender_temp])

In [69]:
df_rating_gender.head()

index    rating  gender
0           318: Shawshank Redemption, The (1994)  0.600000       0
1                        2571: Matrix, The (1999)  0.571429       0
2           593: Silence of the Lambs, The (1991)  0.555556       0
3  260: Star Wars: Episode IV - A New Hope (1977)  0.500000       0
4                    527: Schindler's List (1993)  0.500000       0

In [74]:
df = df_rating_gender.pivot(index='index', columns='gender', values='rating').reset_index()
df['diff'] = df[1]- df[0]
df = df.sort_values(by = 'diff', ascending = False)
df.head()

gender                             index     0         1      diff
6       2396: Shakespeare in Love (1998)  0.00  0.750000  0.750000
13                       34: Babe (1995)  0.00  0.571429  0.571429
17              541: Blade Runner (1982)  0.20  0.750000  0.550000
14              356: Forrest Gump (1994)  0.00  0.500000  0.500000
11              296: Pulp Fiction (1994)  0.25  0.666667  0.416667

In [75]:
df

gender                                              index         0         1  \
6                        2396: Shakespeare in Love (1998)  0.000000  0.750000   
13                                        34: Babe (1995)  0.000000  0.571429   
17                               541: Blade Runner (1982)  0.200000  0.750000   
14                               356: Forrest Gump (1994)  0.000000  0.500000   
11                               296: Pulp Fiction (1994)  0.250000  0.666667   
4                                     1: Toy Story (1995)  0.200000  0.571429   
1       1210: Star Wars: Episode VI - Return of the Je...  0.250000  0.500000   
5                        2028: Saving Private Ryan (1998)  0.285714  0.500000   
12                  318: Shawshank Redemption, The (1994)  0.600000  0.800000   
2                                1259: Stand by Me (1986)  0.250000  0.428571   
8          260: Star Wars: Episode IV - A New Hope (1977)  0.500000  0.571429   
19                     780: Independence Day (ID4) (1996)  0.285714  0.333333   
15                                 3578: Gladiator (2000)  0.500000  0.500000   
9                           2762: Sixth Sense, The (1999)  0.333333  0.333333   
3                              1265: Groundhog Day (1993)  0.500000  0.333333   
10                              2916: Total Recall (1990)  0.200000  0.000000   
18                  593: Silence of the Lambs, The (1991)  0.555556  0.285714   
16                           527: Schindler's List (1993)  0.500000  0.166667   
7                                2571: Matrix, The (1999)  0.571429  0.200000   
0                    1198: Raiders of the Lost Ark (1981)  0.500000  0.000000   

gender      diff  
6       0.750000  
13      0.571429  
17      0.550000  
14      0.500000  
11      0.416667  
4       0.371429  
1       0.250000  
5       0.214286  
12      0.200000  
2       0.178571  
8       0.071429  
19      0.047619  
15      0.000000  
9       0.000000  
3      -0.166667  
10     -0.200000  
18     -0.269841  
16     -0.333333  
7      -0.371429  
0      -0.500000

In [76]:
df.describe()

gender          0          1       diff
count   20.000000  20.000000  20.000000
mean     0.324087   0.438095   0.114008
std      0.195529   0.231347   0.349472
min      0.000000   0.000000  -0.500000
25%      0.200000   0.321429  -0.175000
50%      0.285714   0.500000   0.125000
75%      0.500000   0.571429   0.382738
max      0.600000   0.800000   0.750000